api/envoy/api/v2/core/base.proto

```diff
@@ -76,10 +77,45 @@ message Locality {
   string sub_zone = 3;
 }
 
+// BuildVersion combines SemVer version of extension with free-form build information
+// (i.e. 'alpha', 'private-build') as a set of strings.
+message BuildVersion {
+  // SemVer version of extension
+  type.SemanticVersion version = 1;
+
+  // Free-form build information i.e. 'alpha'
+  repeated string labels = 2;
+}
+
+// Version and identification for an Envoy extension.
+// [#next-free-field: 6]
+message Extension {
+  // This is the name of the Envoy filter as specified in the Envoy
+  // configuration, e.g. envoy.router, com.acme.widget.
+  string name = 1;
+
+  // Category of the extension.
+  // Extension category is provided by the `category()` method of the factory class.
+  // [#comment:TODO(yanavlasov): Link to the doc with existing category names.]
+  string category = 2;
+
+  // Type URL of extension configuration proto.
+  // [#comment:TODO(yanavlasov): Link to the doc with existing configuration protos.]
+  string type_url = 3;
+
+  // The version is a property of the extension and maintained independently
+  // of other extensions and the Envoy API.
+  // This field is not set when extension version is the same as the user agent.
```

Comment:
This is a bit unclear to me still, since build version here is full semantic version, but user agent version is a free form string.

Reply:
I've reworked the user_agent_version to be oneoff of either a free form string or structured BuildVersion.